### PR TITLE
Feat/bn 254

### PR DESCRIPTION
**Pull request by PatStiles**
# Add BN254

## Description

Implementation of BN254 #548  to be finished with @protocolwhisper 

Implements:

- [x] BN254 curve
- [x] Towered Extension Field
- [x] Twist
- [ ] Unit tests
- [ ] Differential Fuzzer against ark_bn254
- [ ] Pairing
- [ ] Optimized Pairing Final Exponentiation as described in https://eprint.iacr.org/2015/192.pdf

## Type of change

Please delete options that are not relevant.

- [x] New feature

## Checklist
- [x] Linked to Github Issue
- [ ] Unit tests added
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
- [ ] This change is an Optimization
  - [ ] Benchmarks added/run

### CODE REVIEW

File: math/src/elliptic_curve/short_weierstrass/curves/bn_254/twist.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    /// This is meant only to be used in the miller loop of the
+    /// ate pairing before the final exponentiation.
+    /// This is because elements in 𝔽p² raised to that
+    /// power are 1 and so the final result of the ate pairing
+    /// doens't depend on having this function output the exact
+    /// values of x' and y'. And it is enough to work with x and y.
+    pub fn to_fp12_unnormalized(&self) -> [FieldElement<Degree12ExtensionField>; 2] {
+        todo!()
+    }
+}
```
Comment (MauroToscano) [inline math/src/elliptic_curve/short_weierstrass/curves/bn_254/twist.rs:77]: This can be removed since it's not implemented

```suggestion
}
```

Comment (MauroToscano) [inline math/src/elliptic_curve/short_weierstrass/curves/bn_254/twist.rs:77]:
```suggestion
```

File: README.md [hunk 2]
```diff
@@ -236,3 +236,5 @@ The following links, repos and projects have been important in the development o
 - [EthSTARK](https://github.com/starkware-libs/ethSTARK/tree/master)
 - [CAIRO whitepaper](https://eprint.iacr.org/2021/1063.pdf)
 - [Gnark](https://github.com/Consensys/gnark)
+
+testing
```
Comment (MauroToscano) [inline README.md:240]:
```suggestion
```

**Review by diegokingston — APPROVED** [review]

File: math/src/elliptic_curve/short_weierstrass/curves/bn_254/twist.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            Level0FE::new(U256::from_hex_unchecked(
+                "a5490e3b00bc8e434f9a1ba734b05c27c525889bf117bb4d293f5aa54b238c5",
+            )),
+            Level0FE::new(U256::from_hex_unchecked(
+                "136ce0ba382e5d37c3e05eff8365e0e6857eefa150096af33bdbdf327649c0eb",
+            )),
+        ]);
+        let p = BN254TwistCurve::create_point_from_affine(px, py).unwrap();
+        let q = BN254TwistCurve::create_point_from_affine(qx, qy).unwrap();
+        //let expected = BN254TwistCurve::create_point_from_affine(expectedx, expectedy).unwrap();
```
Comment (MauroToscano) [inline math/src/elliptic_curve/short_weierstrass/curves/bn_254/twist.rs:163]:
```suggestion
```

**Review by diegokingston — APPROVED** [review]

**Review by MauroToscano — APPROVED** [review]

**Pull request merged**